### PR TITLE
fix asm printer for affine expr (#201)

### LINKED ISSUE #201: affine expression print bug - parenthesis ellision
```
$ cat /tmp/bug.mlir 
#map = (d0)[s0] -> ((-s0 floordiv 4) + 1024)
func @foo(memref<10xf32, #map>)
```
The above is printed with an extra left parenthesis.
```
$ mlir-opt /tmp/bug.mlir
#map0 = (d0)[s0] -> ((-s0 floordiv 4 + 1024)
module {
  func @foo(memref<10xf32, #map0>)
}
```

```
$ mlir-opt /tmp/bug.mlir  | mlir-opt
<stdin>:4:1: error: expected ',' or ')'
module {
^
```

### PR DESCRIPTION
**Pull request by bondhugula**
- fixes #201 and another affine expr print bug

Signed-off-by: Uday Bondhugula <uday@polymagelabs.com>

### CODE REVIEW

File: test/IR/affine-map.mlir [hunk 1]
```diff
@@ -174,6 +174,9 @@
 // CHECK: #map{{[0-9]+}} = (d0) -> (d0 - (d0 + 1))
 #map53 = (d0) -> (d0 - (d0 + 1))
 
+// CHECK: #map{{[0-9]+}} = (d0)[s0] -> ((-s0) floordiv 4, d0 floordiv -1)
+#map54 = (d0)[s0] -> (-s0 floordiv 4, d0 floordiv -1)
```
Comment (ftynse) [inline test/IR/affine-map.mlir]: The RHS of floordiv is expected to be positive according to the spec. I'm assuming that's the reason why the code originally assumed the -1 RHS implied multiplication.  Maybe we should fix the affine verifier instead to disallow negative constants with floordiv/ceildiv/mod.

Comment (bondhugula) [reply test/IR/affine-map.mlir]: > The RHS of floordiv is expected to be positive according to the spec. I'm assuming that's the reason why the code originally assumed the -1 RHS implied multiplication. 

No, that was due to another bug. The change in lib/IR/AsmPrinter.cpp in this commit makes it clear: d0 floordiv -1 or d0 ceildiv -1 was being turned into -d0 due to a missing guard!

> Maybe we should fix the affine verifier instead to disallow negative constants with floordiv/ceildiv/mod.

Negative constants can appear later in the IR as a result of simplification and they could appear on branches that are never taken (basically dead code that is not yet marked dead). As such, they have to be just preserved: this includes floordiv/ceildiv/mod w.r.t zero as well besides negative values. Such expressions are left as is and aren't simplified or folded.

**Comment by bondhugula** [comment]
Anything more on this? All test cases pass for me, and there isn't detail on what's failing internally. @ftynse

**Comment by River707** [comment]
Just fixed it, should be out soon. Apologies for the delay.

**Pull request closed without merging**